Convert admin order cancellation workflow to stimulus reflex

cyrillefr opened the pull request:
#### What? Why?

- Closes #9447 
<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

By
- removing Angular Controller & jQuery calls
- replaces it with Stimulus Controller+ StimulusReflex


#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Visit ... page.
- Visit /admin/orders
Select one or many orders, click on `Actions` then `Cancel Orders` : it should work exactly as before.
- Unit T: `rspec spec/system/admin/bulk_order_cancellation_spec.rb` should raise no failures.


#### Release notes

##### Before
call to Angular Controller
`app/assets/javascripts/admin/orders/controllers/bulk_cancel_controller.js.coffee` that calls `ofnCancelOrderAlert` (jQuery in `variant_autocomplete.js.erb)`

##### After
Call to a ModalConfirmComponent like it is used in #9804
I have updated the original component to take a special message to mimic the behavior of the actual modal template use in the index view of admin/order. So, like the resend confirmation emails, I make use of ViewComponent, to have a similar look/experience.
The call to bulk cancel has been put in its own service, since it is not called anymore as a http get/post. It is now called as a service from a reflex controller.

cyrillefr commented:
NB I was not able to add a flash notice like it is used currently.

I have left a line commented with flash maybe to be sorted later.

The way it works is a page reload after the modifications have taken place (like it works currently).
My guess is the flash value is not taken into account after the reload. Maybe the session object in the reflex controller cannot pass this value to the session.

jibees commented on app/webpacker/controllers/cancel_orders_controller.js:
```diff
@@ -0,0 +1,30 @@
+
+    document
+      .querySelectorAll("#listing_orders input[name='order_ids[]']:checked")
+      .forEach((checkbox) => {
```
Is this possible to use `target` (for `send_cancellation_email`, `restock_items `, ...)  instead of the whole document via `document.querySelector`?
Every time it's possible, I'd prefer this since we are not too strongly tied to DOM structure.

cyrillefr replied:
I have not access to the checkboxes by using target.
Controller is called from within the modal and its scope does not contain checkboxes;

This is why I used a DOM query selector just like you have done in #9804

I understand the need, so I tried to use a new feature of StimulusLS: outlets.
I have played a bit with outlets: I have added an outlet attribute in confirm_modal_component.haml.html.
But it does not work and for now, name of this attribute must be the same as the controller.
From what I have read on the StimulusJS issues, this features is still in development. Furthermore, it would need to be wrapped in a  conditional. 
And least, you have to use an outlet with a CSS selector.

Another idea if you want to keep the modal would be instead of calling a stimulus controller from it, to instead trigger a custom event on the tbody of the lines, and a an attribute with each and every controller needed. (But we might calling many ctrls at once?)
But it is only an idea, I have not tested it and it would mean I could not reuse the code from #9804

jibees replied:
Ok, now I see. That's clear. 
Thanks for letting me know about [outlets](https://stimulus.hotwired.dev/reference/outlets), I didn't know about it. Maybe something I should use soon. 

Maybe for that particular case, it's ok to do some querying with `document.querySelectorAll()`. TBH I'm not convinced about how things (or we?) handle such a thing with StimulusJS. 

Conclusion: I think that's perfect for now! Thanks! 🙏

cyrillefr replied:
Thank you @jibees

I think the stack(StimulusJS/StimulusReflex/ViewComponent/ViewComponentReflex) is new and wide.

You can access to the back in StimulusReflax and also in ViewComponentReflex, so which Strategy is the better, is a bit hard to know. Also when you have to handle both some modal + back-end code can be tricky because you want access both to front-end and back-end :)
I think the way you designed it at first is a good way to factor. The part I had been trying to factor is the bottom lines of the view where are the modals for each choice that cannot be factored for now.

jibees reviewed (commented):
Thanks a lot!

I've view comments, please let me know what you think about it!

Also, I'd add a spec that test the case when many orders are canceled. Would you like to write it?

cyrillefr commented on app/reflexes/cancel_orders_reflex.rb:
```diff
@@ -0,0 +1,9 @@
+  def confirm(params)
+    OrdersBulkCancelService.new(params).call
+    cable_ready.dispatch_event(name: "modal:close")
+    # flash[:success] = Spree.t(:order_updated)
```
Hello jibees,
sorry for the delay
I know, I have "borrowed" at first this piece of code :)

The difference is in this case, there is a reload involved and calling flash means flashing just 0.5 secs before reload.
In Reflex, I guess it is not exactly, the session as in a Rails controller.
And adding a flash in a Reflex session as it seems does not trigger in the next Rails controller call.

My guess is to do something with JS after reload, maybe something that can be shared with other components that behave this way.

jibees replied:
Ok, I didn't know about that reload. Sorry.

jibees reviewed (approved):
@cyrillefr 
I'm a bit lost then: is your PR ready then? Maybe it needs another code review. To me, it's ok like this.

cyrillefr commented:
hello @jibees 

Regarding your last comment, the last thing for me would be the spec you told me about at the top of this review: `Also, I'd add a spec that test the case when many orders are canceled.`
I did add a spec in `bulk_order_cancellation_spec.rb` which tests with 2 orders. If you had that in mind, then I think my PR is ready. If not, please tell me what I am missing.

jibees commented on spec/system/admin/bulk_order_cancellation_spec.rb:
```diff
@@ -29,14 +29,17 @@
       end
 
       # Verify that the orders have a STATE of CANCELLED
+      sleep(1)
```
We tend to avoid this kind of `sleep` in our test. 
I tested, and if I remove the line, the test pass locally. 

We could also test the order model state changed.

That's why I suggest:

```suggestion
within '.reveal-modal' do
        expect {
          find_button("Confirm").click
        }.to change { o1.reload.state }.from('complete').to('canceled')
          .and change { o2.reload.state }.from('complete').to('canceled')
      end

      # Verify that the orders have a STATE of CANCELLED
      expect(page).to have_selector('span', text: 'CANCELLED', count: 2)
```

jibees reviewed (commented):
One small comment, with associated suggestion ;)

cyrillefr commented:
Ok @jibees , we have some problems here :D

Check fails if I do not add some sleep(1) on my spec.
But, this one corrected, I have another problem with `spec/system/admin/orders_spec.rb`.
Which I did not change and do not fails locally. But the big problem is that the code checked is the last version , a version that I do not have locally.
It seems to me CI checks for a version of the code I do not have locally .... which would be a first.
I am even afraid to report it, as it never happened to me before.
Are you able to see if there is anything wrong ?
Maybe I should rebase ?

jibees commented:
> Check fails if I do not add some sleep(1) on my spec.

I tested by removing this `sleep(1)` on line `#45` and run the spec 10 times through:
```
script/rspec-slow-repeat 10  spec/system/admin/bulk_order_cancellation_spec.rb
```
which results: `10 of 10 passed (100%)`

All green! ✅ 
➡️ I think we can remove than line then.

> But [...] I have another problem with spec/system/admin/orders_spec.rb.
Which I did not change and do not fails locally.

I re-triggered a CI run, and the test failed also.
Then I run the spec 10 times locally:
```
script/rspec-slow-repeat 10  spec/system/admin/orders_spec.rb:275
```

Results: `10 of 10 passed (100%)`



> Maybe I should rebase ?

To me, the conclusion is: Yes, I think it's a great idea.

cyrillefr commented:
Hello @jibees , I have rebased.

With regards to the original commits, I have: 
 - modified/squashed spec for orders bulk cancel
 - modified spec for spec_orders
 These should be the last 2 commits

The rest is unchanged

mkllnk reviewed (approved):
Nice work, thank you! :+1:

drummer83 commented:
Hi @cyrillefr,

Thanks for this pull request! I did some testing and here is what I have found.

| Test case | as admin user | as super admin user |
| - | - | - |
| cancel single order in bulk | :heavy_check_mark: | :heavy_check_mark: |
| cancel multiple orders in bulk | :heavy_check_mark: | :heavy_check_mark: |
| cancel order on edit order page (drop down) | :heavy_check_mark: | :heavy_check_mark: |
| resume order on edit order page | :heavy_check_mark: | :heavy_check_mark: |
| cancel resumed order in bulk | :heavy_check_mark: | :heavy_check_mark: |
| test cancellation email setting | :heavy_check_mark: | :heavy_check_mark: |
| test returning items to stock settings | :heavy_check_mark: | :heavy_check_mark: |
| test different order states (pending, ready, shipped; balance due, credit owed, paid, void) | :heavy_check_mark: | :heavy_check_mark: |
| print invoices in bulk | :heavy_check_mark: | :heavy_check_mark: |
| resend confirmation in bulk | :heavy_check_mark: | :heavy_check_mark: |
| compare modals for different bulk actions | :heavy_check_mark: see below | :heavy_check_mark: see below |
| compare loading times to master | :heavy_check_mark: see below | :heavy_check_mark: see below |

Here is the comparison of the two confirmation modals we are using. The are both using the same design. :heavy_check_mark:
![image](https://user-images.githubusercontent.com/1790176/220189048-5cc56820-1da4-4978-8db4-34f9b36f75ed.png)
![image](https://user-images.githubusercontent.com/1790176/220189069-9d1805e2-bc36-4294-af0b-c7e88d3788fd.png)

Here is a little statistics about loading times (tested manually). I had the feeling that loading takes much longer than it used to, so I wanted to check this. But there is no difference. Long loading times are due to the huge amount of data on the server, maybe.
| page | loading time with your PR as admin user | loading time with your PR as super admin | loading time on master (same server) as admin user | loading time on master (same server) as super admin |
| - | - | - | - | - |
| admin/orders | 1.7 s; 1.5 s; 1.5 s; 1.5 s; 1.5 s; :heavy_check_mark: | 3.7 s; 2.6 s; 2.6 s; 2.6 s; 2.7 s :heavy_check_mark: | 3.6 s; 2.0 s; 2.8 s; 1.5 s; 1.8 s; :heavy_check_mark: | 3.4 s; 2.9 s; 3.0 s; 3.2 s; 2.8 s; :heavy_check_mark: |
| /admin/orders/R734522881/edit | 33.5 s; 34.4 s; 34.1 s; 35.0 s :heavy_check_mark: | 37.6 s; 38.1 s; 39.1 s; 38.4 s :heavy_check_mark: | 36.2 s; 35.9 s; 36.5 s; 34.1 s :heavy_check_mark: | 38.9 s; 39.9 s; 39.0 s; 39.7 s; :heavy_check_mark: |

### Result
I couldn't find any problems here. Everything I tested seems to work as before.
Thanks a lot! This one is ready to go! :rocket: :partying_face:
